Fix constant optimization and fused instruction

qwe661234 opened the pull request:
1. Fix tailcall updating in fused instructions
2. Fix errors in constant optimization

Close: #228

jserv commented on src/emulate.c:
```diff
@@ -935,21 +936,22 @@ static void match_pattern(block_t *block)
                 ir->impl = dispatch_table[ir->opcode];
-            } else {
+                ir->tailcall = next_ir->tailcall;
+            } else if (next_ir->opcode == rv_insn_lui) {
```
Is it possible to rewrite with switch-case statements for readability?

jserv commented on src/emulate.c:
```diff
@@ -918,41 +918,40 @@ static void match_pattern(block_t *block)
-            } else {
+            switch (next_ir->opcode) {
+            case rv_insn_add:
+                if (ir->rd == next_ir->rs2) {
```
Can you avoid duplicating code in L923 to L934?

RinHizakura commented on src/rv32_constopt.c:
```diff
@@ -757,7 +756,7 @@ CONSTOPT(csw, {})
 CONSTOPT(caddi, {
     if (constopt_info->is_constant[ir->rd]) {
         constopt_info->is_constant[ir->rd] = true;
```
It may not relate directly in this PR, but I'm Not sure why should we assign `true` to a variable that we have already known as true in the if statement. For example:

 ```cpp
 if (constopt_info->is_constant[ir->rd]) {
        constopt_info->is_constant[ir->rd] = true;
```

It just looks like `constopt_info->is_constant[ir->rd] = true` is a redundant operation. Or do I miss something?

qwe661234 replied:
Yes, it is redundant, I'll remove it.

RinHizakura commented on src/rv32_constopt.c:
```diff
@@ -815,9 +812,9 @@ CONSTOPT(clui, {
+        ir->rd = ir->rs1;
         ir->opcode = rv_insn_clui;
         ir->impl = dispatch_table[ir->opcode];
     } else
```
How about those else cases here? A statement like `constopt_info->is_constant[ir->rs1] = false` here should also be redundant.

qwe661234 replied:
Yes, it is redundant, I'll remove it.

RinHizakura commented:
Since it now looks great on the architectural test, I am fine with the change now. Thanks!

qwe661234 committed:
Fix constant optimization and fused instruction

1. Fix tailcall updating in fused instructions
2. Fix errors in constant optimization

Close: #228
